API/chat/forms.py

```diff
@@ -51,3 +54,67 @@ def save(self, message):
         message.typing = self.cleaned_data.get('typing', False)
 
         message.save()
+
+
+class SessionForm(AuthenticationForm):
+    username = forms.CharField(max_length=20)
+    password = forms.CharField(max_length=32, widget=forms.PasswordInput)
+
+    def clean(self):
+        username = self.cleaned_data.get('username')
+        password = self.cleaned_data.get('password')
+
+        if username is None:
+            raise forms.ValidationError(
+                'invalid_username',
+                code='invalid_username'
+           )
+
+        try:
+            user = User.objects.get(username=username)
+
+            if user.has_usable_password():
+                if password is None:
+                    raise forms.ValidationError(
+                        'password_required',
+                        code='password_required'
+                    )
+
+                auth = authenticate(username=username, password=password)
+
+                if auth is None:
+                    raise forms.ValidationError(
+                        'wrong_password',
+                        code='wrong_password'
+                    )
+
+                return "Authenticate"
+
+            raise forms.ValidationError(
+                'username_reserved',
+                code='username_reserved'
+            )
+
+        except User.DoesNotExist:
+            if password is not None:
+                raise forms.ValidationError(
+                    'password_set',
+                    code='password_set'
```

Comment:
Fill in at `ARCHITECTURE.md`

Reply:
exists already

![image](https://user-images.githubusercontent.com/7916791/37269698-63e10342-25d4-11e8-8186-45ecf157cc19.png)